tests/test_wrappers/test_sklearn_wrapper.py

```diff
@@ -176,17 +175,32 @@ def test_wrap_polynomial_features():
     )
     Xw = tr_wrap.fit_transform(X)
 
-    pd.testing.assert_frame_equal(Xw, pd.concat([X, Xt], axis=1))
-    assert Xw.shape[1] == len(X.columns) + len(tr.get_feature_names_out(varlist))
+    pd.testing.assert_frame_equal(Xw, pd.concat([X.drop(columns=varlist), Xt], axis=1))
+    assert Xw.shape[1] == len(X.drop(columns=varlist).columns) + len(
+        tr.get_feature_names_out(varlist)
+    )
 
     # when variable list is None
     tr_wrap.set_params(variables=None)
 
     Xt = pd.DataFrame(tr.fit_transform(X), columns=tr.get_feature_names_out())
     Xw = tr_wrap.fit_transform(X)
 
-    pd.testing.assert_frame_equal(Xw, pd.concat([X, Xt], axis=1))
-    assert Xw.shape[1] == len(X.columns) + len(tr.get_feature_names_out(X.columns))
+    pd.testing.assert_frame_equal(Xw, Xt)
+    assert Xw.shape[1] == len(tr.get_feature_names_out(X.columns))
+
+
+def test_wrap_polynomial_features_get_features_name_out():
+    X = fetch_california_housing(as_frame=True).frame
+
+    tr = PolynomialFeatures()
+    tr_wrap = SklearnTransformerWrapper(transformer=PolynomialFeatures())
+    varlist = ["MedInc", "HouseAge", "AveRooms", "AveBedrms"]
+
+    tr.fit(X[varlist])
+    tr_wrap.fit(X[varlist])
+
+    assert (tr.get_feature_names_out() == tr_wrap.get_feature_names_out()).all()
```

Comment:
This is the test for Polynomial Features

Reply:
the test that is missing is something like this:

```
def test_wrap_polynomial_features_get_features_name_out():
    X = fetch_california_housing(as_frame=True).frame

    varlist = ["MedInc", "HouseAge", "AveRooms", "AveBedrms"]
    tr_wrap = SklearnTransformerWrapper(transformer=PolynomialFeatures(), variables=varlist)
   
    tr_wrap.fit(X)

    assert tr_wrap.get_feature_names_out() == The expected list (all features in the original + new features)
    assert tr_wrap.get_feature_names_out(varlist) == All the new polynomial features
    assert tr_wrap.get_feature_names_out(["Medinc") == the features expected from Medinc
```

The thing is, given that the logic in get_feature_names_out was changed slightly, I would like to make sure that it still returns what it is expected to return given the input to input_features

Would you be able to change it?

Reply:
Sure. I think the assertion I have covers your first assertion. I can add the remaining two. Or do you want me to hard code the original features + new features in the first one?

Reply:
@solegalli, I was checking this one, and just to make sure I didn't touch `get_feature_names_out`. The thing is I'm not sure what is the expected behavior for this one: `assert tr_wrap.get_feature_names_out(["Medinc") == the features expected from Medinc`.
This one should work even if it was trained on a different pool of features? And the expected output is: Medinc, Medinc^2 and all the interactions that includes Medinc?

Reply:
Assuming that the transformer was trained as in my example:

```
    varlist = ["MedInc", "HouseAge", "AveRooms", "AveBedrms"]
    tr_wrap = SklearnTransformerWrapper(transformer=PolynomialFeatures(), variables=varlist)
```

Then the output of tr_wrap.get_feature_names_out(["Medinc") would be as you say the polynomial combinations that involve MedInc. And this, should come out of the box from the Polynomial Features.

If the transformer was trained like this instead:

```
    varlist = ["HouseAge", "AveRooms", "AveBedrms"]
    tr_wrap = SklearnTransformerWrapper(transformer=PolynomialFeatures(), variables=varlist)
```

where MedInc was not part of variables, then the outcome of tr_wrap.get_feature_names_out(["Medinc") should be an error, and I think this should be handled by our class method get_feature_names_out.

Reply:
@solegalli, This issue was a bit more complicated than I thought. Actually I noticed the following things:

* First, the current implementation throws an error because as per Sklearn Documentation `input_features` needs to be equals to `name_features_in` when using `get_feature_names_out` in `PolynomialFeatures` and `OneHotEncoder`. See this:

![image](https://user-images.githubusercontent.com/48638337/185290240-15590cbf-ff4c-4031-842f-93ece4f410b9.png)
 * So adding something like `tr_wrap.get_feature_names_out(["Medinc"])` is not valid when trained with more features.

* This means current implementation does not work as expected and cannot be obtained out of the box as mentioned previously. 
 
I implemented a solution when input_features is not None, but I find it not very elegant:

```python
# Get the names of all the new features
added_features = self.transformer_.get_feature_names_out(
                    self.variables_
                )
# Get all the features related to input_features... Sorry for the double for loop
feature_names = []
for feature in added_features:
    for _in in input_features:
        if _in in feature and feature not in feature_names:
            feature_names.append(feature)

# In case of PolynomialFeatures and include_bias is True I need to also retrieve the Bias called '1'
if (
    self.transformer_.__class__.__name__ == "PolynomialFeatures"
    and self.transformer_.include_bias
):
    feature_names = ["1"] + feature_names
```

The reason why I'm using this double for loop (specially for PolynomialFeatures)  is because I need to check all the created features (squared ones, interactions, etc.) related to an input features are included and not repeated. Adittionally in case of include bias, I'm adding '1' as part of the output feature names, I did this because it was the expected behavior for `test_get_feature_names_out_polynomialfeature`. After a lot of trial an error (breaking a lot of tests) I came up to this solution, but I'm totally open to make it better. Didn't find a better way to implement this with lists. I tried sets, but they don't preserve the order of features, so order of features is a bit messy and difficult to test.

The thing is, if an input_variable is not part of the training variables not error is thrown, so what error should I raise if this happens?

PS: I'm currently passing all the tests and I'm getting the results as expected in our discussion, if you accept a solution like this I can push my changes.

Thanks!!

Reply:
Hi @datacubeR 

Thank you so much for so much detail.

I wonder what the point was in offering input_features as parameter in the PolynomialFeatures , if then you can only pass feature_names_in :/

In this case, I think we do not need to modify the code further or add the additional test.

Sorry, that was my bad. For some reason I thought that you could pass one variable and obtain the derived features.